File: functional/tinyweb/src/main/java/com/lamronby/oo/tinyweb/TinyWeb.java
package com.lamronby.oo.tinyweb;

import java.util.List;
import java.util.Map;

public class TinyWeb {
	private Map<String, Controller> controllers;
	private List<Filter> filters;

	public TinyWeb(Map<String, Controller> controllers, List<Filter> filters) {
		this.controllers = controllers;
		this.filters = filters;
	}

	public HttpResponse handleRequest(HttpRequest request) {

		HttpRequest currentRequest = request;
		for (Filter filter : filters) {
			currentRequest = filter.doFilter(request);
		}

		Controller controller = controllers.get(request.getPath());

		if (null == controller)
			return null;

		return controller.handleRequest(request);
	}
}

File: functional/tinyweb/src/main/scala/com/lamronby/oo/tinyweb/stepfour/TinyWeb.scala
package com.lamronby.oo.tinyweb.stepfour

class TinyWeb(controllers: Map[String, Controller],
	filters: List[(HttpRequest) => HttpRequest]) {


	def handleRequest(request: HttpRequest): Option[HttpResponse] = {
		val composedFilter = filters.reverse.reduceLeft(
			(composed, next) => composed compose next)
		val filteredRequest = composedFilter(httpRequest)
		val controllerOption = controllers.get(filteredReuqest.path)
		controllerOption amp { controller => controller.handleRequest(filteredRequest) }
	}
}